js/blocks/controls/fetch-input.js

```diff
@@ -0,0 +1,304 @@
+const { withInstanceId } = wp.compose;
+const apiFetch = wp.apiFetch;
+const { addQueryArgs } = wp.url;
+
+// Since FetchInput is rendered in the context of other inputs, but should be
+// considered a separate modal node, prevent keyboard events from propagating
+// as being considered from the input.
+const stopEventPropagation = ( event ) => event.stopPropagation();
+
+class FetchInput extends Component {
+	constructor( { autocompleteRef } ) {
+		super( ...arguments );
+
+		this.onBlur = this.onBlur.bind( this );
+		this.onChange = this.onChange.bind( this );
+		this.onKeyDown = this.onKeyDown.bind( this );
+		this.autocompleteRef = autocompleteRef || createRef();
+		this.inputRef = createRef();
+		this.updateSuggestions = this.updateSuggestions.bind( this );
+
+		this.suggestionNodes = [];
+
+		this.state = {
+			results: [],
+			showSuggestions: false,
+			selectedSuggestion: null,
+		};
+	}
+
+	componentWillUnmount() {
+		delete this.suggestionsRequest;
+	}
+
+	bindSuggestionNode( index ) {
+		return ( ref ) => {
+			this.suggestionNodes[ index ] = ref;
+		};
+	}
+
+	updateSuggestions( value ) {
+		// Show the suggestions after typing at least 2 characters
+		// and also for URLs.
+		if ( value.length < 2 ) {
+			this.setState( {
+				showSuggestions: false,
+				selectedSuggestion: null,
+				loading: false,
+			} );
+
+			return;
+		}
+
+		this.setState( {
+			showSuggestions: true,
+			selectedSuggestion: null,
+			loading: true,
+		} );
+
+		const request = apiFetch( {
+			path: addQueryArgs( '/wp/v2/' + this.props.apiSlug, {
+				search: value,
+				per_page: 20,
+			} ),
+		} );
+
+		request.then( ( results ) => {
+			// A fetch Promise doesn't have an abort option. It's mimicked by
+			// comparing the request reference in on the instance, which is
+			// reset or deleted on subsequent requests or unmounting.
+			if ( this.suggestionsRequest !== request ) {
+				return;
+			}
+
+			this.setState( {
+				results,
+				loading: false,
+			} );
+
+			if ( !! results.length ) {
+				this.props.debouncedSpeak( sprintf( _n(
+					'%d result found, use up and down arrow keys to navigate.',
+					'%d results found, use up and down arrow keys to navigate.',
+					results.length,
+					'block-lab'
+				), results.length ), 'assertive' );
+			} else {
+				this.props.debouncedSpeak( __( 'No results.', 'block-lab' ), 'assertive' );
+			}
+		} ).catch( () => {
+			if ( this.suggestionsRequest === request ) {
+				this.setState( {
+					loading: false,
+				} );
+			}
+		} );
+
+		this.suggestionsRequest = request;
+	}
+
+	/**
+	 * On clicking the block outside the <input>, hide the Popover.
+	 * Mainly taken from the color control onBlur handler.
+	 */
+	onBlur( event ) {
+		console.log( event.relatedTarget.classList );
+		if (
+			event.relatedTarget
+			&&
+			(
+				event.relatedTarget.classList.contains( 'wp-block' ) // The block editor.
+				||
+				event.relatedTarget.classList.contains( 'edit-post-sidebar' ) // The inspector.
+			)
+		) {
```

Comment:
I just realised that this doesn't work if the new target is another input inside the same block. For example, consider a block with a user field and a text field. Type a few search characters in the user field, see the popover, then select the text field. The expected behaviour is that the popover would disappear, but it doesn't.

I remember this being a really tough problem to solve. If you feel that you've spent too long on it, feel free to just leave this for now and create a new issue to address it. It would be good if there were a way, though.

Reply:
Hi @lukecarbis, good catch. https://github.com/getblocklab/block-lab/pull/233/commits/1b6c13b68fd38f2e4de66423c5a666499c56af07 looks to fix this locally:

![text-field-clicking](https://user-images.githubusercontent.com/4063887/54194007-e03b5c00-4480-11e9-95db-0d6f5a2f4178.gif)